File: 74/search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int M = matrix.size(), N = matrix[0].size();
        int left=0, right = M*N;
        while(left<right){
            int mid = (left+right)/2;
            printf("left:%d mid:%d right:%d\n", left, mid, right);
            int i = mid/N, j = mid%N;
            if(matrix[i][j]==target) return true; //小心
            if(matrix[i][j]<target) left = mid+1;
            else right = mid;
        }
        return false;
/*        printf("left:%d\n", left);
        if(left<M*N && matrix[left/N][left%N]==target) return true;
        else return false;*/
    }
};
//case 3/133: [[1]] 1
//case 4/133: 

File: 74/search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        M, N = len(matrix), len(matrix[0])
        # 直接用 binary search 就可以了
        left, right = 0, M*N
        while left < right:
            print(left, right)
            mid = int((left+right)/2)
            i, j = int(mid/N), int(mid%N)
            now = matrix[i][j]
            # print("matrix[i][j]:", now)
            if now==target: return True;
            if now > target: right = mid
            else: left = mid + 1
        
        return False
        
